### PR TITLE
Shenli/wrap data

### PR DESCRIPTION
**Pull request by shenli**
Wrap data with FieldType for cast function.
Fix https://github.com/pingcap/tidb/issues/325

In the next pr, I will change the parameters of compare/cast/convert from raw type to DataItem.

### CODE REVIEW

File: util/types/field_type.go [hunk 1]
```diff
@@ -111,3 +111,13 @@ func (ft *FieldType) String() string {
 
 	return strings.Join(strs, " ")
 }
+
+// DataItem is wrapped data with type info.
```
Comment (qiuyesuifeng) [inline util/types/field_type.go]: is wrapped -> warps?

File: plan/plans/union.go [hunk 2]
```diff
@@ -19,6 +19,7 @@ import (
 	"github.com/pingcap/tidb/expression"
 	"github.com/pingcap/tidb/field"
 	"github.com/pingcap/tidb/kv/memkv"
+	mysql "github.com/pingcap/tidb/mysqldef"
```
Comment (qiuyesuifeng) [inline plan/plans/union.go]: Need merge master.

File: plan/plans/fields.go [hunk 3]
```diff
@@ -78,10 +78,19 @@ func (r *SelectFieldsDefaultPlan) Next(ctx context.Context) (row *plan.Row, err
...
 	for i, fld := range r.Fields {
-		var err error
-		if row.Data[i], err = fld.Expr.Eval(ctx, r.evalArgs); err != nil {
+		d, err := fld.Expr.Eval(ctx, r.evalArgs)
+		if err != nil {
 			return nil, errors.Trace(err)
 		}
+		di, ok := d.(*types.DataItem)
+		if ok {
+			if r.ResultFields[i].Col.Tp == 0 {
```
Comment (qiuyesuifeng) [inline plan/plans/fields.go]: Can 0 use a const value?

File: util/types/field_type.go (hunk 1 above)
Comment (shenli) [inline util/types/field_type.go]: wrapped is an adj.

File: plan/plans/fields.go (hunk 3 above)
Comment (shenli) [inline plan/plans/fields.go]: PTAL @qiuyesuifeng

**Comment by shenli** [comment]
PTAL @qiuyesuifeng

**Pull request merged**